vnext/JSI/Shared/PreparedJavaScriptStore.h

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+namespace facebook::jsi {
+class PreparedJavaScript;
+}
+
+namespace Microsoft::JSI {
+
+/**
+ * TODO (yicyao)
+ */
+class PreparedJavaScriptStore {
```

Comment:
>PreparedJavaScriptStore [](start = 6, length = 23)

I would appreciate some design feedback/suggestions here. Here are the problems that I find with the architecture proposed in this PR:

1. For now, let's assume that a user of ReactNative passes a RuntimeHolderLazyInit (see RuntimeHolder.h) as part of DevSettings. Consider a sample implementation of ChakraCoreJsiHolder. In order to construct a ChakraCoreRuntime, one need to be able to construct a ChakraCoreRuntimeArgs. In order to construct a ChakraCoreRuntimeArgs, one need to be able to construct a PreparedJavaScriptStore. A PreparedJavaScriptStore for ChakraCore will need to know about ChakraCorePreparedJavaScript. This means that we need to expose ChakraCoreJsiHolder, ChakraCoreRuntimeArgs, and ChakraCorePreparedJavaScript to the user of react native. This seems just bad.

2. Since every engine is different, each engine's PreparedJavaScript will likely be different, and for each engine, we might want different PreparedJavaScriptStore that cache/read PreparedJavaScript in different ways. We run into this combinatorial explosion. I think breaking this interface up into something that handles caching and reading and something that handles interpreting the structure of PreparedJavaScript may help with this. I am not sure where version checking would fit here though.

Reply:
For #2, I agree that it feels like there are two components. Things like writing bytecode to disk would be shared between different runtimes, but generating information about things like versioning or creating the actual bytecode would be runtime specific.

For versioning, you could have the runtime give you something like a hash. E.g. for Chakra, you could combine all the normal fields about version with a Chakra specific magic number or something.

Maybe you could add an interface like `PreparedScriptWriter` that could output a buffee and runtime version?

---
In reply to: [345402558](https://github.com/microsoft/react-native-windows/pull/3650#discussion_r345402558) [](ancestors = 345402558)

Reply:
#1 looks ok. If Engine supports various options, they should all be exposed to the user. 

---
In reply to: [345472382](https://github.com/microsoft/react-native-windows/pull/3650#discussion_r345472382) [](ancestors = 345472382,345402558)